File: python/11. container-with-most-water.py
from typing import List


# lc sol start
class Solution:
    @staticmethod
    def get_area(height: List[int], start_idx: int, end_idx: int) -> int:
        return (end_idx - start_idx) * min(height[start_idx], height[end_idx])

    def maxArea(self, height: List[int]) -> int:
        start_idx, end_idx = 0, len(height) - 1
        max_area = float('-inf')

        while start_idx < end_idx:
            max_area = max(max_area, self.get_area(height, start_idx, end_idx))

            if height[start_idx] < height[end_idx]:
                start_idx += 1
            else:
                end_idx -= 1

        return max_area


# lc sol end


def main() -> None:
    test_cases = {
        'inputs': [([1, 8, 6, 2, 5, 4, 8, 3, 7], ), ([1, 1], )],
        'outputs': [49, 1]
    }

    solution = Solution()

    for idx, args in enumerate(test_cases['inputs']):
        op = solution.maxArea(*args)
        assert op == test_cases['outputs'][idx], \
            f"expected output {test_cases['outputs'][idx]} for input {args}, got {op}"

    print('all sample test cases passed successfully!')


if __name__ == '__main__':
    main()

File: cpp/11. container-with-most-water.cpp
//
// Created by Rutvik Patel on 11/18/23.
//

#include "include.h"

// solution

class Solution {
private:
    int get_area(std::vector<int>& height, int start_idx, int end_idx) {
        return (end_idx - start_idx) * std::min(height[start_idx], height[end_idx]);
    }
public:
    int maxArea(std::vector<int>& height) {
        int start_idx = 0, end_idx = height.size() - 1;
        int max_area = get_area(height, start_idx, end_idx);

        while (start_idx < end_idx) {
            max_area = std::max(get_area(height, start_idx, end_idx), max_area);

            if (height[start_idx] < height[end_idx])
                start_idx++;
            else
                end_idx--;
        }

        return max_area;
    }
};

// main

int main(__attribute__((unused)) int argc, __attribute__((unused)) char *argv[]) {
    std::vector<std::pair<std::vector<int>, int>> test_cases = {
        { { 1, 1 }, 1 },
        { { 1, 8, 6, 2, 5, 4, 8, 3, 7 }, 49 },
    };

    Solution solution;

    for (auto [heights, capacity] : test_cases)
        assert_m(solution.maxArea(heights) == capacity, "Failed!");

    log("All test cases passed!");
}